### ISSUE TITLE
Backup dashboard with API and Import Dashboard in UI

### ISSUE THREAD

**Issue by rp-happiestbaby** [grafana/grafana#9152]
Hi,
I made backup using Grafana rest API and curl,
curl -s -k -H "Authorization: Bearer $key" $grafana_host/api/dashboards/db/dashboard-name 
I can see JSON and it looks ok all vars are there 

Now I want to copy and past JSON to UI and import dashboard 
It asked me to enter name (didn't asked to select data source) but after import dashboard is completely empty.

UI and API dashboard import/export are not compatible?

 
Grafana version : v4.4.3 (commit: 54c79c5)
datasource : Graphite
OS : Ubuntu 14.04

**Comment by torkelo** [comment]
if you want to get/create dashboards via api use the normal dashboard save/update api:
http://docs.grafana.org/http_api/dashboard/#create-update-dashboard

The UI export/import feature is for exporting / importing between instances, and includes data source info. 

You say after import the dashboard was empty? empty how? empty of panels? empty of data? any errors? did you check the panel data sources, are they set to a data source you have?

**Comment by rp-happiestbaby** [comment]
@torkelo thanks for your response.
So I was exporting via API and importing via GUI (And after Import dashboard was completely empty)

Is there any way to export dashboard via API(Or curl) for exporting / importing between instances and include data source info? 
I need to use curl and export dashboard the way that GUI export is doing.

**Comment by torkelo** [comment]
as long as the data source name is the same it should work. works when I test and have no other reports on empty dashboard after importing a dashboard via UI that has been exported via normal GET dashboard API. should work fine as long as data source names match (or default data source match)

**Comment by rp-happiestbaby** [comment]
I have one simple dashboard(One text box with html `<h1>Test</h1>`) that no datasource Is required 

i=dashboard_name
key=my_grafana_key
grafana_host=my_grafana_host

curl -s -k -H "Authorization: Bearer $key" $grafana_host/api/dashboards/db/$i > ${i}.json

This is content of file 
`{"meta":{"type":"db","canSave":false,"canEdit":false,"canStar":true,"slug":"d2","expires":"0001-01-01T00:00:00Z","created":"2017-09-06T22:11:40Z","updated":"2017-09-06T22:11:40Z","updatedBy":"admin","createdBy":"admin","version":1},"dashboard":{"dashboard":{"annotations":{"list":[]},"editable":true,"gnetId":null,"graphTooltip":0,"hideControls":false,"id":null,"links":[],"rows":[{"collapse":false,"height":"250px","panels":[{"content":"\u003ch1\u003eTest\u003c/h1\u003e","id":1,"mode":"html","span":12,"title":"Panel Title","type":"text"}],"repeat":null,"repeatIteration":null,"repeatRowId":null,"showTitle":false,"title":"Dashboard Row","titleSize":"h6"}],"schemaVersion":14,"style":"dark","tags":[],"templating":{"list":[]},"time":{"from":"now-6h","to":"now"},"timepicker":{"refresh_intervals":["5s","10s","30s","1m","5m","15m","30m","1h","2h","1d"],"time_options":["5m","15m","1h","6h","12h","24h","2d","7d","30d"]},"timezone":"","title":"D2","version":1},"id":32,"meta":{"canEdit":false,"canSave":false,"canStar":true,"created":"2017-09-06T22:08:44Z","createdBy":"admin","expires":"0001-01-01T00:00:00Z","slug":"d2","type":"db","updated":"2017-09-06T22:08:44Z","updatedBy":"admin","version":1},"overwrite":true,"title":"D2","version":1}}
`

When I trying import this dashboard via GUI Dashboard is empty.

**Comment by torkelo** [comment]
copy paste only the dashboard part of the response

**Comment by torkelo** [comment]
this is the dashboard json:

```
{
  "annotations": {
    "list": [
      
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "hideControls": false,
  "id": null,
  "links": [
    
  ],
  "rows": [
    {
      "collapse": false,
      "height": "250px",
      "panels": [
        {
          "content": "<h1>Test<\/h1>",
          "id": 1,
          "mode": "html",
          "span": 12,
          "title": "Panel Title",
          "type": "text"
        }
      ],
      "repeat": null,
      "repeatIteration": null,
      "repeatRowId": null,
      "showTitle": false,
      "title": "Dashboard Row",
      "titleSize": "h6"
    }
  ],
  "schemaVersion": 14,
  "style": "dark",
  "tags": [
    
  ],
  "templating": {
    "list": [
      
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {
    "refresh_intervals": [
      "5s",
      "10s",
      "30s",
      "1m",
      "5m",
      "15m",
      "30m",
      "1h",
      "2h",
      "1d"
    ],
    "time_options": [
      "5m",
      "15m",
      "1h",
      "6h",
      "12h",
      "24h",
      "2d",
      "7d",
      "30d"
    ]
  },
  "timezone": "",
  "title": "D2",
  "version": 1
}
```

**Comment by rp-happiestbaby** [comment]
I need to backup dashboard and have JSON the way that UI Export dose it,
Is there any way to do it with curl ?

**Comment by imatpro** [comment]
Hi, I have a same problem, do you have resolv it?

**Comment by violenti** [comment]
HI @torkelo, I also need export dashboard from api and do import from UI. Is it possible? 
For example,  a dashboard  obtain from the api, when I import from UI, is empty the new dashboard:
```
{
  "meta": {
    "type": "db",
    "canSave": true,
    "canEdit": true,
    "canAdmin": false,
    "canStar": true,
    "slug": "ci-and-cd",
    "url": "/d/000000030/ci-and-cd",
    "expires": "0001-01-01T00:00:00Z",
    "created": "2018-04-11T20:13:02Z",
    "updated": "2018-04-12T01:30:28Z",
    "updatedBy": "xxxxx",
    "createdBy": "xxxx",
    "version": 2,
    "hasAcl": false,
    "isFolder": false,
    "folderId": 0,
    "folderTitle": "General",
    "folderUrl": ""
  },
  "dashboard": {
    "annotations": {
      "list": [
        {
          "builtIn": 1,
          "datasource": "-- Grafana --",
          "enable": true,
          "hide": true,
          "iconColor": "rgba(0, 211, 255, 1)",
          "name": "Annotations & Alerts",
          "type": "dashboard"
        }
      ]
    },
    "editable": true,
    "gnetId": null,
    "graphTooltip": 0,
    "hideControls": false,
    "id": null,
    "links": [],
    "rows": [
      {
        "collapse": false,
        "height": 333,
        "panels": [
          {
            "aliasColors": {},
            "bars": false,
            "dashLength": 10,
            "dashes": false,
            "datasource": "Prometheus",
            "fill": 1,
            "id": 1,
            "legend": {
              "avg": false,
              "current": false,
              "max": false,
              "min": false,
              "show": true,
              "total": false,
              "values": false
            },
            "lines": true,
            "linewidth": 1,
            "links": [],
            "nullPointMode": "null",
            "percentage": false,
            "pointradius": 5,
            "points": false,
            "renderer": "flot",
            "seriesOverrides": [],
            "spaceLength": 10,
            "span": 6,
            "stack": false,
            "steppedLine": false,
            "targets": [
              {
                "expr": "sum(container_memory_usage_bytes{pod_name=~\"ci-jenkins-.*\"} / 1024 /1024  )  by (namespace, pod_name)\n",
                "format": "time_series",
                "intervalFactor": 2,
                "refId": "A"
              }
            ],
            "thresholds": [],
            "timeFrom": null,
            "timeShift": null,
            "title": "Memory jenkins",
            "tooltip": {
              "shared": true,
              "sort": 0,
              "value_type": "individual"
            },
            "type": "graph",
            "xaxis": {
              "buckets": null,
              "mode": "time",
              "name": null,
              "show": true,
              "values": []
            },
            "yaxes": [
              {
                "format": "decmbytes",
                "label": null,
                "logBase": 1,
                "max": null,
                "min": null,
                "show": true
              },
              {
                "format": "decmbytes",
                "label": null,
                "logBase": 1,
                "max": null,
                "min": null,
                "show": true
              }
            ]
          },
          {
            "aliasColors": {},
            "bars": false,
            "dashLength": 10,
            "dashes": false,
            "datasource": "Prometheus",
            "fill": 1,
            "id": 3,
            "legend": {
              "avg": false,
              "current": false,
              "max": false,
              "min": false,
              "show": true,
              "total": false,
              "values": false
            },
            "lines": true,
            "linewidth": 1,
            "links": [],
            "nullPointMode": "null",
            "percentage": false,
            "pointradius": 5,
            "points": false,
            "renderer": "flot",
            "seriesOverrides": [],
            "spaceLength": 10,
            "span": 6,
            "stack": false,
            "steppedLine": false,
            "targets": [
              {
                "expr": "sum(container_memory_usage_bytes{pod_name=~\"cd-spinnaker-.*\"} / 1024 /1024  )  by (namespace, pod_name)\n",
                "format": "time_series",
                "intervalFactor": 2,
                "refId": "A"
              }
            ],
            "thresholds": [],
            "timeFrom": null,
            "timeShift": null,
            "title": "Memory cd-spinnaker",
            "tooltip": {
              "shared": true,
              "sort": 0,
              "value_type": "individual"
            },
            "type": "graph",
            "xaxis": {
              "buckets": null,
              "mode": "time",
              "name": null,
              "show": true,
              "values": []
            },
            "yaxes": [
              {
                "format": "decmbytes",
                "label": null,
                "logBase": 1,
                "max": null,
                "min": null,
                "show": true
              },
              {
                "format": "decmbytes",
                "label": null,
                "logBase": 1,
                "max": null,
                "min": null,
                "show": true
              }
            ]
          }
        ],
        "repeat": null,
        "repeatIteration": null,
        "repeatRowId": null,
        "showTitle": false,
        "title": "Dashboard Row",
        "titleSize": "h6"
      },
      {
        "collapse": false,
        "height": 312,
        "panels": [
          {
            "aliasColors": {},
            "bars": false,
            "dashLength": 10,
            "dashes": false,
            "datasource": "Prometheus",
            "fill": 1,
            "id": 2,
            "legend": {
              "avg": false,
              "current": false,
              "max": false,
              "min": false,
              "show": true,
              "total": false,
              "values": false
            },
            "lines": true,
            "linewidth": 1,
            "links": [],
            "nullPointMode": "null",
            "percentage": false,
            "pointradius": 5,
            "points": false,
            "renderer": "flot",
            "seriesOverrides": [],
            "spaceLength": 10,
            "span": 6,
            "stack": false,
            "steppedLine": false,
            "targets": [
              {
                "expr": "sum(rate(container_cpu_usage_seconds_total{pod_name=~\"ci-jenkins-.*\"}[5m])* 1000) by (namespace, pod_name)",
                "format": "time_series",
                "intervalFactor": 2,
                "refId": "A"
              }
            ],
            "thresholds": [],
            "timeFrom": null,
            "timeShift": null,
            "title": "CPU Jenkins",
            "tooltip": {
              "shared": true,
              "sort": 0,
              "value_type": "individual"
            },
            "type": "graph",
            "xaxis": {
              "buckets": null,
              "mode": "time",
              "name": null,
              "show": true,
              "values": []
            },
            "yaxes": [
              {
                "format": "short",
                "label": null,
                "logBase": 1,
                "max": null,
                "min": null,
                "show": true
              },
              {
                "format": "short",
                "label": null,
                "logBase": 1,
                "max": null,
                "min": null,
                "show": true
              }
            ]
          },
          {
            "aliasColors": {},
            "bars": false,
            "dashLength": 10,
            "dashes": false,
            "datasource": "Prometheus",
            "fill": 1,
            "id": 4,
            "legend": {
              "avg": false,
              "current": false,
              "max": false,
              "min": false,
              "show": true,
              "total": false,
              "values": false
            },
            "lines": true,
            "linewidth": 1,
            "links": [],
            "nullPointMode": "null",
            "percentage": false,
            "pointradius": 5,
            "points": false,
            "renderer": "flot",
            "seriesOverrides": [],
            "spaceLength": 10,
            "span": 6,
            "stack": false,
            "steppedLine": false,
            "targets": [
              {
                "expr": "sum(rate(container_cpu_usage_seconds_total{pod_name=~\"cd-spinnaker-.*\"}[5m])* 1000) by (namespace, pod_name)",
                "format": "time_series",
                "intervalFactor": 2,
                "refId": "A"
              }
            ],
            "thresholds": [],
            "timeFrom": null,
            "timeShift": null,
            "title": "CPU cd-spinnaker",
            "tooltip": {
              "shared": true,
              "sort": 0,
              "value_type": "individual"
            },
            "type": "graph",
            "xaxis": {
              "buckets": null,
              "mode": "time",
              "name": null,
              "show": true,
              "values": []
            },
            "yaxes": [
              {
                "format": "short",
                "label": null,
                "logBase": 1,
                "max": null,
                "min": null,
                "show": true
              },
              {
                "format": "short",
                "label": null,
                "logBase": 1,
                "max": null,
                "min": null,
                "show": true
              }
            ]
          }
        ],
        "repeat": null,
        "repeatIteration": null,
        "repeatRowId": null,
        "showTitle": false,
        "title": "Dashboard Row",
        "titleSize": "h6"
      },
      {
        "collapse": false,
        "height": 250,
        "panels": [],
        "repeat": null,
        "repeatIteration": null,
        "repeatRowId": null,
        "showTitle": false,
        "title": "Dashboard Row",
        "titleSize": "h6"
      }
    ],
    "schemaVersion": 14,
    "style": "dark",
    "tags": [],
    "templating": {
      "list": []
    },
    "time": {
      "from": "now-6h",
      "to": "now"
    },
    "timepicker": {
      "refresh_intervals": [
        "5s",
        "10s",
        "30s",
        "1m",
        "5m",
        "15m",
        "30m",
        "1h",
        "2h",
        "1d"
      ],
      "time_options": [
        "5m",
        "15m",
        "1h",
        "6h",
        "12h",
        "24h",
        "2d",
        "7d",
        "30d"
      ]
    },
    "timezone": "",
    "title": "CI & CD",
    "uid": "000000030",
    "version": 2
  },
  "overwrite": true
}
```
Do you know   what  it can be?  thanks!

**Comment by torkelo** [comment]
Just paste everything inside the dashboard json property in the import UI text box and it will work

**Comment by evanstucker-hates-2fa** [comment]
```
#!/bin/bash

# Usage:
#
# export_grafana_dashboards.sh https://admin:REDACTED@grafana.dedevsecops.com

create_slug () {
  echo "$1" | iconv -t ascii//TRANSLIT | sed -r s/[^a-zA-Z0-9]+/-/g | sed -r s/^-+\|-+$//g | tr A-Z a-z
}

full_url=$1
username=$(echo "${full_url}" | cut -d/ -f 3 | cut -d: -f 1)
base_url=$(echo "${full_url}" | cut -d@ -f 2)
folder=$(create_slug "${username}-${base_url}")

mkdir "${folder}" 
for db_uid in $(curl -s "${full_url}/api/search" | jq -r .[].uid); do
  db_json=$(curl -s "${full_url}/api/dashboards/uid/${db_uid}")
  db_slug=$(echo "${db_json}" | jq -r .meta.slug)
  db_title=$(echo "${db_json}" | jq -r .dashboard.title)
  filename="${folder}/${db_slug}.json"
  echo "Exporting \"${db_title}\" to \"${filename}\"..." 
  echo "${db_json}" | jq -r . > "${filename}"
done
echo "Done"
```

**Comment by keshavmass0** [comment]
@torkelo I am facing similar situation, backing using API and try restoring from UI console. I am also getting empty dash board. Is there any way to get the solution. Kindly confirm

@rp-happiestbaby is solution suggested by torkelo working for you, please confirm as its not working for me

**Comment by danatinflux** [comment]
@keshavmass0 I think I figured it out. 

Originally, my backup script looked something like this:

```
curl -s -u $GPASS -X GET "${URL}/api/dashboards/uid/${DASHID}" | jq -r > ${FILEPATH}/${HOST}/dashboards/${ORGNAME}/${DASHNAME}.json
```

I noticed, like you did, that the results wouldn't copy-paste cleanly, leaving me with a dashboard with no contents. 

So, I did some experimenting with JQ, and it appears that it will work just fine if you strip out the meta info at the top. This can be accomplished by changing the backup command to this:

```
curl -s -u $GPASS -X GET "${URL}/api/dashboards/uid/${DASHID}" | jq -r .dashboard > ${FILEPATH}/${HOST}/dashboards/${ORGNAME}/${DASHNAME}.json
```

Starting the json at the `.dashboard` point appears to take out any extra json and allow an easy copy-paste.

Hope that helps anyone in the same situation!
